### PR TITLE
HTTP Digest Authorization

### PR DESCRIPTION
**Pull request by ondratu**

### CODE REVIEW

File: poorwsgi/digest.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            '{hash1}:{nonce}:{hash2}'
+            ''.format(**kwargs).encode()).hexdigest()
+    return response == kwargs['response']
+
+
+def check_credentials(req, realm, username=None):
+    """Check Digest authorization credentials.
+
+    Return True if Authorization header is valid for realm and username
+    if is set.
```
Comment (uzak) [inline poorwsgi/digest.py:75]: `if is set` ... it feels like something is missing here

File: poorwsgi/digest.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        log.error('Digest: response not match')
+        return False
+    return True
+
+
+def check_digest(realm, username=None):
+    """Check HTTP Digest Authenticate.
+
+    Allow only valid HTTP Digest authorization for realm and username
+    if is set. When no, HTTP_UNAUTHORIZED response was raised with realm
```
Comment (uzak) [inline poorwsgi/digest.py:131]: again, the "if is set" part of the sentence seems to lack something.

File: poorwsgi/digest.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    password = getpass('New password: ')
+    re_type = getpass('Re-type new password: ')
+    if password != re_type:
+        print('They don\'t match, sorry.')
+        return None
+    return password
+
+
+def main():
+    """Main fce for manipulation with passwordfile."""
```
Comment (uzak) [inline poorwsgi/digest.py:245]: fce -> function. (Personally I know only the `fct` abbreviation).

File: poorwsgi/digest.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    # pylint: disable=too-many-statements
+    # pylint: disable=too-many-branches
+
+    parser = ArgumentParser(
+        description="Password file manipulation for digest authenticate"
+                    " method.")
+
+    parser.add_argument(
+        "passwordfile", type=str, nargs='?',
+        help="password file which will be process")
```
Comment (uzak) [inline poorwsgi/digest.py:256]: process -> processed

File: poorwsgi/request.py [hunk 5]
```diff
@@ -266,53 +268,61 @@ def __init__(self, environ, app_config):
...
 
         # variables for user use
         self.__user = None
         self.__api = None
     # enddef
 
     # -------------------------- Properties --------------------------- #
+    @property
+    def app(self):
+        """Return Application object which was create Request."""
```
Comment (uzak) [inline poorwsgi/request.py:323]: create -> created. (like `process -> processed`)

File: poorwsgi/request.py [hunk 6]
```diff
@@ -356,9 +366,25 @@ def method_number(self):
 
     @property
     def uri(self):
-        """The path portion of the URI."""
+        """Deprecated alias for path path of the URI."""
```
Comment (uzak) [inline poorwsgi/request.py:369]: path path

File: poorwsgi/session.py [hunk 7]
```diff
@@ -47,6 +51,43 @@ def hidden(text, passwd):
...
+    """Create token from secret, and client string.
+
+    If timeout is set, token contains time align with twice of this value.
+    Twice, because time of creating can be so near to computed time.
+    """
+    if timeout is None:
+        text = "%s%s" % (secret, client)
+    else:
+        if expired == 0:
+            now = int(time() / timeout) * timeout   # shift to start time
```
Comment (uzak) [inline poorwsgi/session.py:64]: why not just `int(time())`?

File: poorwsgi/session.py [hunk 8]
```diff
@@ -47,6 +51,43 @@ def hidden(text, passwd):
...
+            expired = now + 2 * timeout
+        text = "%s%s%s" % (secret, expired, client)
+
+    return sha256(text.encode()).hexdigest()
+
+
+def check_token(token, secret, client, timeout=None):
+    """Check token, if is right.
```
Comment (uzak) [inline poorwsgi/session.py:72]: if **it** is right (a sentence always need a subject).

File: poorwsgi/session.py [hunk 9]
```diff
@@ -47,6 +51,43 @@ def hidden(text, passwd):
...
+
+    Arguments secret, client and expired must be same, when token was be
+    generated. If expired is set, than token must be younger than 2*expired.
```
Comment (uzak) [inline poorwsgi/session.py:75]: `was be` cannot be said in English. `Will be` (if you want future) or `was` (if you want past).

File: poorwsgi/wsgi.py [hunk 10]
```diff
@@ -393,6 +412,91 @@ def json_mime_types(self):
...
+        # for Digest
+        if self.__config['secret_key'] is None:
+            raise ValueError('Set secret key first')
+        self.__config['auth_type'] = value
+
+    @property
+    def auth_algorithm(self):
+        """Authorization algorithm.
+
+        Algorithm is depend on authorization type and client support.
```
Comment (uzak) [inline poorwsgi/wsgi.py:437]: is depend -> `depends` (or `is depending`)

**Review by uzak — COMMENTED** [review]
It might seem like a lot of language corrections, but I think there is improvement in regard to the past. Well done!

File: doc/documentation.rst [hunk 11]
```diff
@@ -856,6 +856,7 @@ Next to it there are some Request properties, to get parset header values.
                     header.
 :is_xhr:            True if ``X-Requested-With`` is ``XMLHttpRequest``.
 :cookies:           Cooike object created from ``Cookie`` header or empty tuple.
+:authorization:     Parsed ``Authorization`` header to dictionary.
```
[the remark this answers is not in the document]
Comment (ondratu) [reply doc/documentation.rst:859]: Yes, `req.authorization` is dictionary, there are many parameters in header value, much more than in cookie.

File: doc/documentation.rst [hunk 12]
```diff
@@ -946,6 +947,86 @@ hidden function.
...
+~~~~~~~~~~~~~~~~
+
+PoorWSGI support HTTP Digest Authorization from version 2.3.x.
+Supported are:
+
+    * MD5, MD5-sess, SHA-256, SHA-256-sess algorithm, **MD5-sess** is default
+    * none or auth quality of protection (qop), **auth** is default
+    * nonce value timeout, so new hash will be count every N seconds,
+      **300** sec (5min) is default
+    * nc browser value **is not check** on server side now
```
[the remark this answers is not in the document]
Comment (ondratu) [reply doc/documentation.rst:960]: nc is value name, i will use some formatting characters.

File: poorwsgi/digest.py (hunk 1 above)
Comment (ondratu) [reply poorwsgi/digest.py:75]: username is not required.

File: poorwsgi/session.py (hunk 7 above)
Comment (ondratu) [reply poorwsgi/session.py:64]: It must be shifted to timeout period.

File: doc/documentation.rst [hunk 13]
```diff
@@ -946,6 +947,86 @@ hidden function.
...
+
+
+Application settings
+````````````````````
+
+There are some application options, which is used for HTTP Authorization
+configuration.
```
[the remark this answers is not in the document]
Comment (ondratu) [reply doc/documentation.rst:967]: i need your explaining.

File: poorwsgi/digest.py [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+
+from poorwsgi import state
+from poorwsgi.response import HTTPException
+from poorwsgi.session import check_token
+
+log = getLogger("poorwsgi")  # pylint: disable=invalid-name
+
+
+def check_response(req, password):
+    """Check digest response value username/password in fact.
```
[the remark this answers is not in the document]
Comment (ondratu) [reply poorwsgi/digest.py:31]: I need to consult this.

File: poorwsgi/digest.py (hunk 3 above)
Comment (ondratu) [reply poorwsgi/digest.py:245]: fce is from czech language :D

**Pull request merged**